### PR TITLE
Add a workaround for threadpool deadlocks on recursion

### PR DESCRIPTION
**Pull request by nigredo-tori**
Partial solution for https://github.com/nim-lang/Nim/issues/4597 .
Only solves the third case (recursion with cpu load). However it still deadlocks if all the slave threads are blocked in recursion and the pool is full. What should be done in this situation?

### CODE REVIEW

File: lib/pure/concurrency/threadpool.nim [hunk 1]
```diff
@@ -464,10 +476,30 @@ proc nimSpawn3(fn: WorkerProc; data: pointer) {.compilerProc.} =
         fn(self, data)
         await(self.taskStarted)
         return
+
+    if isSlave and numSlavesRunning <= numSlavesWaiting + 1:
```
Comment (Araq) [inline lib/pure/concurrency/threadpool.nim]: seems to be racy. No barriers?

**Comment by Araq** [comment]
> However it still deadlocks if all the slave threads are blocked in recursion and the pool is full. 

What does "blocked in recursion" mean?

> What should be done in this situation?

The spawing thread should run the task directly instead of spawning it, I think.

**Comment by nigredo-tori** [comment]
>> However it still deadlocks if all the slave threads are blocked in recursion and the pool is full. 

> What does "blocked in recursion" mean? 

I mean the slave thread is blocked in recursive spawn call. Waiting on `gSomeReady`. Which will not be triggered by any other slave thread, if all of them are waiting on it. 

>> What should be done in this situation?

> The spawing thread should run the task directly instead of spawning it, I think.

There is something along these lines in code (nimRecursiveSpawn). The problem with running tasks recursively is the following: suppose you recursively execute task that never finishes. This means that the calling task will never finish as well.
Of course, we can document this behavior ("make sure all submitted tasks terminate"), but this still excludes a lot of otherwise valid use cases.

By the way, is there a motivation for this blocking spawn implementation? Wouldn't it be better to just have a task queue?

**Comment by nigredo-tori** [comment]
Added a lock to prevent the race (thanks to @Araq). The code got uglier though (had to hold the lock from condition check to variable increment).

**Comment by Araq** [comment]
> By the way, what is the motivation behind this blocking spawn implementation? Wouldn't it be better to just have a task queue?

Yes indeed that would be better. The motivation was two-fold:

a) I wanted to see if it can work.
b) It saves a copy since we directly reconstruct the task in the thread's heap. Ordinary queue is 2 copies: data -> queue -> data

**Pull request merged**